pkg/services/ngalert/models/testing.go

```diff
@@ -20,320 +19,477 @@ import (
+			if !exists {
 				rule.ID = id
 				return
 			}
+			id = g.Rand.Int63n(1500) + 1
 		}
 	}
 }
 
-func WithGroupIndex(groupIndex int) AlertRuleMutator {
-	return func(rule *AlertRule) {
+func (a *AlertRuleMutators) WithGroupIndex(groupIndex int) AlertRuleMutator {
+	return func(_ *AlertRuleGenerator, rule *AlertRule) {
 		rule.RuleGroupIndex = groupIndex
 	}
 }
 
-func WithUniqueGroupIndex() AlertRuleMutator {
-	usedIdx := make(map[int]struct{})
-	return func(rule *AlertRule) {
+func (a *AlertRuleMutators) WithUniqueGroupIndex() AlertRuleMutator {
+	usedIdx := sync.Map{}
+	return func(g *AlertRuleGenerator, rule *AlertRule) {
+		idx := rule.RuleGroupIndex
 		for {
-			idx := rand.Int()
-			if _, ok := usedIdx[idx]; !ok {
-				usedIdx[idx] = struct{}{}
+			if _, exists := usedIdx.LoadOrStore(idx, struct{}{}); !exists {
 				rule.RuleGroupIndex = idx
 				return
 			}
+			idx = g.Rand.Int()
 		}
 	}
 }
 
-func WithSequentialGroupIndex() AlertRuleMutator {
+func (a *AlertRuleMutators) WithSequentialGroupIndex() AlertRuleMutator {
 	idx := 1
-	return func(rule *AlertRule) {
+	m := sync.Mutex{}
+	return func(_ *AlertRuleGenerator, rule *AlertRule) {
+		m.Lock()
+		defer m.Unlock()
 		rule.RuleGroupIndex = idx
 		idx++
 	}
 }
 
-func WithOrgID(orgId int64) AlertRuleMutator {
-	return func(rule *AlertRule) {
+func (a *AlertRuleMutators) WithOrgID(orgId int64) AlertRuleMutator {
+	return func(_ *AlertRuleGenerator, rule *AlertRule) {
 		rule.OrgID = orgId
 	}
 }
 
-func WithUniqueOrgID() AlertRuleMutator {
-	orgs := map[int64]struct{}{}
-	return func(rule *AlertRule) {
-		var orgID int64
+func (a *AlertRuleMutators) WithUniqueOrgID() AlertRuleMutator {
+	orgs := sync.Map{}
+	return func(g *AlertRuleGenerator, rule *AlertRule) {
+		orgID := rule.OrgID
 		for {
-			orgID = rand.Int63()
-			if _, ok := orgs[orgID]; !ok {
-				break
+			if _, exist := orgs.LoadOrStore(orgID, struct{}{}); !exist {
+				rule.OrgID = orgID
+				return
 			}
+			orgID = g.Rand.Int63()
 		}
-		orgs[orgID] = struct{}{}
-		rule.OrgID = orgID
 	}
 }
 
 // WithNamespaceUIDNotIn generates a random namespace UID if it is among excluded
-func WithNamespaceUIDNotIn(exclude ...string) AlertRuleMutator {
-	return func(rule *AlertRule) {
+func (a *AlertRuleMutators) WithNamespaceUIDNotIn(exclude ...string) AlertRuleMutator {
+	return func(g *AlertRuleGenerator, rule *AlertRule) {
 		for {
 			if !slices.Contains(exclude, rule.NamespaceUID) {
 				return
 			}
-			rule.NamespaceUID = uuid.NewString()
+			rule.NamespaceUID = util.GenerateShortUID()
 		}
 	}
 }
 
-func WithNamespace(namespace *folder.Folder) AlertRuleMutator {
-	return func(rule *AlertRule) {
+func (a *AlertRuleMutators) WithNamespaceUID(namespaceUID string) AlertRuleMutator {
+	return func(_ *AlertRuleGenerator, rule *AlertRule) {
+		rule.NamespaceUID = namespaceUID
+	}
+}
+
+func (a *AlertRuleMutators) WithNamespace(namespace *folder.Folder) AlertRuleMutator {
+	return func(_ *AlertRuleGenerator, rule *AlertRule) {
```

Comment:
This could technically just return `WithNamespaceUID(namespace.UID)` couldn't it?